### PR TITLE
Updated NodeRef to clientv1.ObjectReference.

### LINKED ISSUE #40979: kube-proxy failed to record events
**Is this a BUG REPORT or FEATURE REQUEST? (choose one):**
bug

**Kubernetes version (use kubectl version):**
master

**What happened:**

Found error again in latest code:

```
E0203 06:40:04.508992       1 event.go:259] Could not construct reference to: '&v1.ObjectReference{Kind:"Node", Namespace:"", Name:"k8s-master", UID:"k8s-master", APIVersion:"", ResourceVersion:"", FieldPath:""}' due to: 'object does not implement the List interfaces'. Will not report event: 'Normal' 'Starting' 'Starting kube-proxy.'
```

**What you expected to happen:**

No errors

**How to reproduce it (as minimally and precisely as possible):**

```
# ./hack/local-up-cluster.sh
# grep "^E" /tmp/kube-proxy.log  | head
```

cc @kubernetes/sig-network-misc

### PR DESCRIPTION
**Pull request by k82cn**
fixes #40979

### CODE REVIEW

**Comment by k82cn** [comment]
The record was updated to `client-go` but did not update `config.NodeRef`; so `GetReference` can not get its object.

/cc @andrewrynhard

**Comment by k82cn** [comment]
Append test result:

```
# kc get event -s http://9.111.143.202:8080
LASTSEEN   FIRSTSEEN   COUNT     NAME        KIND      SUBOBJECT   TYPE      REASON                    SOURCE                  MESSAGE
2m         2m          1         127.0.0.1   Node                  Normal    Starting                  kube-proxy, 127.0.0.1   Starting kube-proxy.
2m         2m          1         127.0.0.1   Node                  Normal    Starting                  kubelet, 127.0.0.1      Starting kubelet.
...
```

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Comment by k82cn** [comment]
cc @deads2k / @smarterclayton

**Comment by deads2k** [comment]
I think you missed this: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L1013

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -203,6 +206,12 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 		glog.Fatalf("Invalid API configuration: %v", err)
 	}
 
+	eventClientConfig := *kubeconfig
```
Comment (deads2k) [inline cmd/kube-proxy/app/server.go]: no need to copy.

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -203,6 +206,12 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
...
+	eventClient, err := clientgoclientset.NewForConfig(&eventClientConfig)
+	if err != nil {
+		glog.Warningf("Failed to create API Server client: %v", err)
```
Comment (deads2k) [inline cmd/kube-proxy/app/server.go]: return the error.  This shouldn't happen unless something pretty bad happened.

File: cmd/kubemark/hollow-node.go [hunk 3]
```diff
@@ -140,7 +141,22 @@ func main() {
 		endpointsConfig := proxyconfig.NewEndpointsConfig()
 		endpointsConfig.RegisterHandler(&kubemark.FakeProxyHandler{})
 
-		hollowProxy := kubemark.NewHollowProxyOrDie(config.NodeName, internalClientset, endpointsConfig, serviceConfig, iptInterface, eventBroadcaster, recorder)
+		eventClientConfig := *clientConfig
```
Comment (deads2k) [inline cmd/kubemark/hollow-node.go]: same comment about copy and error.

**Comment by deads2k** [comment]
One missed call site and a few minor comments.  looks good otherwise.

File: cmd/kube-proxy/app/server.go [hunk 4]
```diff
@@ -203,6 +206,11 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 		glog.Fatalf("Invalid API configuration: %v", err)
 	}
 
+	eventClient, err := clientgoclientset.NewForConfig(kubeconfig)
+	if err != nil {
+		glog.Fatalf("Failed to create API Server client: %v", err)
```
Comment (deads2k) [inline cmd/kube-proxy/app/server.go]: `return nil, err` as above please

**Comment by deads2k** [comment]
Just the error handling, then lgtm.

**Comment by k82cn** [comment]
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmd/kube-proxy/app/server.go, line 209 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40980#-KcDGL1QesoXuQ6FJ_gU-r1-209:-KcIfUj4eTy9hQnupd42:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/b07c70a7d8dfabf3f370a7988807f387ad703dfa/cmd/kube-proxy/app/server.go#L209)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

no need to copy.
</blockquote></details>

Done.

---

*[cmd/kube-proxy/app/server.go, line 212 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40980#-KcDGL1QesoXuQ6FJ_gU-r1-212:-KcIfVah4C-0BbVf51t0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/b07c70a7d8dfabf3f370a7988807f387ad703dfa/cmd/kube-proxy/app/server.go#L212)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

return the error.  This shouldn't happen unless something pretty bad happened.
</blockquote></details>

Done.

---

*[cmd/kube-proxy/app/server.go, line 211 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40980#-KcDGL1QesoXuQ6FJ_gU-r2-211:-KcIfU9aGTyjhouuOdR3:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/d504e7e635d30cc4264aa404d6fe32806bc705cd/cmd/kube-proxy/app/server.go#L211)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

`return nil, err` as above please
</blockquote></details>

Done.

---

*[cmd/kubemark/hollow-node.go, line 144 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40980#-KcDV8oTgcbra7avH4vi-r1-144:-KcIfWdN7bDKX3ksFgRc:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/b07c70a7d8dfabf3f370a7988807f387ad703dfa/cmd/kubemark/hollow-node.go#L144)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary><blockquote>

same comment about copy and error.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/40980)*

**Comment by deads2k** [comment]
approved based on https://github.com/kubernetes/kubernetes/issues/40363

**Comment by k82cn** [comment]
@k8s-bot gci gce e2e test this

**Comment by deads2k** [comment]
@eparis something change with the bot and tags?

**Comment by deads2k** [comment]
huh, how much you want to bet that his clock is in the future.

**Comment by apelisse** [comment]
@deads2k Nothing! that's exactly what's going on ;-)

**Comment by shyamjvs** [comment]
@k82cn lgtm for the kubemark part (assuming that it works based on the kubemark e2e test)

**Comment by deads2k** [comment]
@k82cn you should check your system clock.  Don't re-push, but somehow github thinks you're about a day ahead of us.

**Comment by k82cn** [comment]
> @k82cn you should check your system clock. Don't re-push, but somehow github thinks you're about a day ahead of us.

Sure, let me check it.

**Pull request merged**